### PR TITLE
Consolidate edn peg rules to better parse keywords and symbols

### PR DESCRIPTION
**Pull request by victorporof**
Fixes https://github.com/mozilla/mentat/issues/219

### CODE REVIEW

**Comment by victorporof** [comment]
I've also updated peg to 0.5.1 from 0.4 in this PR. Second commit is relevant to #219

**Comment by ncalexan** [comment]
> I've also updated peg to 0.5.1 from 0.4 in this PR. Second commit is relevant to #219

Next time, can you say why you're upgrading?  There are lots of changes in the changelog, did you need a particular feature?  Or are you just trying to stay fresh?

File: edn/src/edn.rustpeg [hunk 1]
```diff
@@ -76,25 +76,22 @@ namespace_separator = "/"
 symbol_char_initial = [a-z] / [A-Z] / [0-9] / [*!_?$%&=<>]
 symbol_char_subsequent = [a-z] / [A-Z] / [0-9] / [-*!_?$%&=<>]
 
-symbol_namespace = symbol_char_initial+ (namespace_divider symbol_char_subsequent+)*
+symbol_namespace = symbol_char_initial+ symbol_char_subsequent* (namespace_divider symbol_char_subsequent+)*
```
Comment (ncalexan) [inline edn/src/edn.rustpeg]: Shouldn't this be `= symbol_char_initial symbol_char_subsequent`, with no `+`?

File: edn/tests/tests.rs [hunk 2]
```diff
@@ -113,9 +113,16 @@ fn test_symbol() {
...
+    assert_eq!(keyword(":foo-bar/baz-boz").unwrap(), k_ns("foo-bar", "baz-boz"));
+
+    assert_eq!(keyword(":foo-bar/baz_boz").unwrap(), k_ns("foo-bar", "baz_boz"));
+    assert_eq!(keyword(":foo_bar/baz-boz").unwrap(), k_ns("foo_bar", "baz-boz"));
+    assert_eq!(keyword(":foo_bar/baz_boz").unwrap(), k_ns("foo_bar", "baz_boz"));
 
     assert_eq!(keyword(":symbol").unwrap(), k_plain("symbol"));
     assert_eq!(keyword(":hello").unwrap(), k_plain("hello"));
+    assert_eq!(keyword(":foo-bar").unwrap(), k_plain("foo-bar"));
+    assert_eq!(keyword(":foo_bar").unwrap(), k_plain("foo_bar"));
```
Comment (ncalexan) [inline edn/tests/tests.rs]: Very nice!  Consider duplicating for symbols, just to document our expectations.

Can you also test that empty namespace and/or empty name is rejected?

File: edn/src/edn.rustpeg [hunk 3]
```diff
@@ -76,25 +76,22 @@ namespace_separator = "/"
...
-
 pub symbol -> Value =
-    ns:( sns:$(symbol_namespace) namespace_separator { sns })? n:$(symbol_name) {
+    ns:( sns:$(symbol_namespace) namespace_separator {
+        sns
+    })? n:$(symbol_name) {
         types::to_symbol(ns, n)
     }
 
 pub keyword -> Value =
```
Comment (ncalexan) [inline edn/src/edn.rustpeg]: Might be worth commenting that the character rules for keywords are the same as for symbols, per https://github.com/edn-format/edn#keywords.

**Comment by victorporof** [comment]
I'll get to landing this asap tomorrow morning.

**Comment by victorporof** [comment]
> Next time, can you say why you're upgrading? 
Wanted to stay fresh and upgrading process seemed trivial/automatic enough to not feel like wasting time.

File: edn/src/edn.rustpeg (hunk 1 above)
Comment (victorporof) [inline edn/src/edn.rustpeg]: Sure!

Comment (rnewman) [inline edn/src/edn.rustpeg]: It definitely needs `*` for symbol_char_subsequent, because the next thing in the production is a namespace divider.

It doesn't matter whether the initial `symbol_char_initial` is a `+` or not, because `s_c_i` is a subset of `s_c_s`.

Comment (ncalexan) [inline edn/src/edn.rustpeg]: Fuck, that was one fix and one typo.  I want to the standard "one initial, zero-or-more subsequent" pattern.

File: edn/src/edn.rustpeg (hunk 3 above)
Comment (victorporof) [inline edn/src/edn.rustpeg]: I've read that, and the following stood out:

> Keywords follow the rules of symbols, except they can (and must) begin with :, e.g. :fred or :my/fred

It's the reason I unified those regexps. Am I misunderstanding the above? Should we not do this?

File: edn/tests/tests.rs (hunk 2 above)
Comment (victorporof) [inline edn/tests/tests.rs]: Will do

File: edn/src/edn.rustpeg (hunk 3 above)
Comment (ncalexan) [inline edn/src/edn.rustpeg]: No, you're correct -- I just think we might call it out in the code, that keywords follow the same rules as symbol (module the leading `:`).

File: edn/src/edn.rustpeg (hunk 1 above)
Comment (victorporof) [inline edn/src/edn.rustpeg]: > Shouldn't this be = symbol_char_initial symbol_char_subsequent, with no +?

Yup, my brain intuitively read that as `symbol_char_initial symbol_char_subsequent*`.

**Comment by victorporof** [comment]
Addressed comments.
